group-coordinator/src/main/java/org/apache/kafka/coordinator/group/metrics/HdrHistogram.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.metrics;
+
+import org.HdrHistogram.Histogram;
+import org.HdrHistogram.Recorder;
+import org.HdrHistogram.ValueRecorder;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * <p>A wrapper on top of the HdrHistogram API. It handles writing to the histogram by delegating
+ * to an internal {@link ValueRecorder} implementation, and reading from the histogram by
+ * efficiently implementing the retrieval of up-to-date histogram data.
+ *
+ * <p>Note that all APIs expect a timestamp which is used by the histogram to discard decaying data
+ * and determine when the snapshot from which the histogram metrics are calculated should be
+ * refreshed.
+ */
+public final class HdrHistogram {
```

Comment:
Historically, we have been using the histogram in Yammer metric on the server side. Is there a particular reason that we want to introduce HdrHistogram? It would be use to add some comments to guide future developers on which histogram metric to use.

Reply:
hey @junrao, my understanding is that we are trying to move away from yammer and the community is looking into alternative solutions at the moment, for instance an OpenTelemetry based histogram. In the meantime we are planning to use this implementation and the original author @dimitarndimitrov will be moving this to server-common to be used as the server histogram.

Reply:
@jeffkbkim: Yes, we do plan to move away from yammer in the long term. But it seems that we should add HdrHistogram in org.apache.kafka.common.metrics.Metrics so that both the client and the server can use the same histogram. We probably need a KIP for that.

Reply:
> and the original author @dimitarndimitrov will be moving this to server-common to be used as the server histogram

@jeffkbkim This is not exactly correct. There is no certainty around a move to `server-common`, because the aforementioned looking for alternatives is still ongoing.

Reply:
@dimitarndimitrov : Thanks for the reply. Are you saying that the usage of HdrHistogram is still experimental and not intended from broader use?

Reply:
> Are you saying that the usage of HdrHistogram is still experimental and not intended from broader use?

@junrao Not really, I think with the fix I've mentioned in another comment, this will be practically the same as an internal `HdrHistogram` implementation that's fairly well proven in production usage by now.

I only wanted to underscore that we can't directly proclaim this as the main preferred histogram implementation in AK without coordinating with the other related efforts that we're aware of, which focus on an OpenTelemetry-based implementation.

Reply:
> Is HdrHistogram better than Yammer histogram?

@junrao I think we have good evidence that it is.
The included unit tests illustrate a bit how HdrHistogram is more accurate than the Yammer histogram, and JMH benchmarks point to significantly faster writes (I have concrete numbers from internal tests, but I'll try to include a JMH benchmark in the bugfix PR I'm preparing, and we can look at the results it gives us).

> If it's better, why not use it more broadly on the server side? What's our recommendation for new developers who want to add histogram metrics on the server side?

I personally would vouch for that. Part of the motivation for originally implementing this was seamlessly replacing the Yammer histogram usages in `RequestMetrics`, which can be noticeable in profiles in high-load scenarios, and using a more performant histogram could be a nice little win in such cases.

All I'm saying is that we don't claim to have convinced the community already, and don't want to inadvertently undermine the other "new histogram" effort we were made aware of, before there's been a discussion.

Reply:
> All I'm saying is that we don't claim to have convinced the community already, and don't want to inadvertently undermine the other "new histogram" effort we were made aware of, before there's been a discussion.

Ok. What's our plan moving forward? We probably don't want to be in the interim state for too long.

Reply:
> What's our plan moving forward? We probably don't want to be in the interim state for too long.

Agreed. I'll try to sync with the other party in the next weeks and provide an update.

Reply:
I agree with Jun that it's odd to take a dependency on a library unless we're confident it's the way forward long-term. Let's work towards a resolution before the next release please.